protobuf/decode.go

```diff
@@ -0,0 +1,282 @@
+		if err != nil {
+			return nil, fmt.Errorf("Error decoding entry account; %w", err)
+		}
+
+		code, err := lookup(codehash)
+		if err != nil && !errors.Is(err, leveldb.ErrNotFound) {
+			return nil, fmt.Errorf("Error looking up codehash; %w", err)
+		}
+
+		world[address] = substate.NewAccount(nonce, balance, code)
+		for _, storage := range acct.GetStorage() {
+			key, value, err := storage.decode()
+			if err != nil {
+				return nil, fmt.Errorf("Error decoding account storage entry; %w", err)
+			}
+			world[address].Storage[key] = value
+		}
+	}
+
+	return &world, nil
+}
+
+func (entry *Substate_AllocEntry) decode() ([]byte, *Substate_Account, error) {
+	return entry.GetAddress(), entry.GetAccount(), nil
+}
+
+func (acct *Substate_Account) decode() (uint64, *big.Int, []byte, types.Hash, error) {
+	return acct.GetNonce(),
+		types.BytesToBigInt(acct.GetBalance()),
+		acct.GetCode(),
+		types.BytesToHash(acct.GetCodeHash()),
+		nil
+}
+
+func (entry *Substate_Account_StorageEntry) decode() (types.Hash, types.Hash, error) {
+	return types.BytesToHash(entry.GetKey()),
+		types.BytesToHash(entry.GetValue()),
+		nil
+}
+
+// decode converts protobuf-encoded Substate_BlockEnv into aida-comprehensible Env
+func (env *Substate_BlockEnv) decode() (*substate.Env, error) {
+	var blockHashes map[uint64]types.Hash = nil
+	if env.GetBlockHashes() != nil {
+		blockHashes = make(map[uint64]types.Hash, len(env.GetBlockHashes()))
+		for _, entry := range env.GetBlockHashes() {
+			key, value, err := entry.decode()
+			if err != nil {
+				return nil, err
+			}
+			blockHashes[key] = types.BytesToHash(value)
+		}
+	}
+
+	return &substate.Env{
+		Coinbase:    types.BytesToAddress(env.GetCoinbase()),
+		Difficulty:  types.BytesToBigInt(env.GetDifficulty()),
+		GasLimit:    env.GetGasLimit(),
+		Number:      env.GetNumber(),
+		Timestamp:   env.GetTimestamp(),
+		BlockHashes: blockHashes,
+		BaseFee:     BytesValueToBigInt(env.GetBaseFee()),
+		Random:      BytesValueToHash(env.GetRandom()),
+		BlobBaseFee: BytesValueToBigInt(env.GetBlobBaseFee()),
+	}, nil
+}
+
+func (entry *Substate_BlockEnv_BlockHashEntry) decode() (uint64, []byte, error) {
+	return entry.GetKey(), entry.GetValue(), nil
+}
+
+// decode converts protobuf-encoded Substate_TxMessage into aida-comprehensible Message
+func (msg *Substate_TxMessage) decode(lookup dbGetCode) (*substate.Message, error) {
+
+	// to=nil means contract creation
+	var pTo *types.Address = nil
+	to := msg.GetTo()
+	if to != nil {
+		address := types.BytesToAddress(to.GetValue())
+		pTo = &address
+	}
+
+	// if InitCodeHash exists:
+	// 1. code = lookup the code using InitCodeHash
+	// 2. set data -> code from (1)
+	// 3. clear InitCodeHash
+	var data []byte = msg.GetData()
+	if pTo == nil {
+		code, err := lookup(types.BytesToHash(msg.GetInitCodeHash()))
+		if err != nil && !errors.Is(err, leveldb.ErrNotFound) {
+			return nil, fmt.Errorf("failed to decode tx message; %w", err)
+		}
+		data = code
+	}
```

Comment:
this is strange piece of code, this is how protobuf should work?
Also you are not doing number 3. - I dont see where you clear it

Reply:
Not sure which line exactly are you referring to.
Regarding 3, protobuf does this for us automatically through `oneof`. (or at least the auto-generated protobuf code should do that for us). That's why I put that as a comment.

https://github.com/Fantom-foundation/Substate/blob/87cfa91562d8f842b53fc325d0237f8afa4de26a/protobuf/substate.proto#L68-L71

Reply:
The whole condition seems odd - first you extract data from msg. Then if `pTo` is nil you use the `code` as data. Is that what we want?

Reply:
Yes, data defaults to `msg.GetData()`, unless `pTo=nil && msg.GetInitCodeHash() != nil`. 
The latter is the case where the tx is trying to create a contract.

Note that `data` is not necessarily code when not creating a new contract.

Reply:
That being said, I think this warrants a cleaner alternative to communicate all of this.

Reply:
[ced5262](https://github.com/Fantom-foundation/Substate/pull/85/commits/ced5262f98e5843a3d625ac34d9415106cba8c5d)